src/librustc_passes/ast_validation.rs

```diff
@@ -202,19 +209,21 @@ impl<'a> Visitor<'a> for AstValidator<'a> {
 
     fn visit_item(&mut self, item: &'a Item) {
         match item.node {
-            ItemKind::Impl(.., Some(..), _, ref impl_items) => {
+            ItemKind::Impl(.., ref generics, Some(..), _, ref impl_items) => {
                 self.invalid_visibility(&item.vis, item.span, None);
                 for impl_item in impl_items {
                     self.invalid_visibility(&impl_item.vis, impl_item.span, None);
                     if let ImplItemKind::Method(ref sig, _) = impl_item.node {
                         self.check_trait_fn_not_const(sig.constness);
                     }
                 }
+                generics.lifetimes.iter().for_each(|l| self.check_lifetime(&l.lifetime))
```

Comment:
It's not necessary to check lifetimes in all possible contexts manually, you can add a `visit_lifetime` method to `impl<'a> Visitor<'a> for AstValidator<'a>` and call `check_lifetime` from it.

`Visitor` is kind of an object-oriented infrastructure for walking through AST. The default visitor in `libsyntax/visit.rs` is a base visitor that does nothing and only walks the tree and other concrete visitors (like `AstValidator`) can override its methods and add something useful to them.

Reply:
This is actually a good first task because the visitor is a quite important piece of infrastructure and it makes you familiar with it.